roles/foreman_image/templates/Dockerfile.j2

```diff
@@ -0,0 +1,38 @@
+FROM centos:7
+
+RUN echo "tsflags=nodocs" >> /etc/yum.conf && \
+    yum -y install http://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm \
+                   http://yum.theforeman.org/releases/{{ version }}/el7/x86_64/foreman-release.rpm && \
+    yum -y install foreman-release-scl && \
+    yum -y install foreman foreman-postgresql tfm-rubygem-puma tfm-rubygem-redis && yum clean all
+
+# Generate apipie:cache:index
+RUN yum -y install tfm-rubygem-activerecord-nulldb-adapter && yum clean all
+
+COPY container-assets/schema.rb.nulldb /usr/share/foreman/db/schema.rb
+COPY container-assets/nulldb.rb /usr/share/foreman/bundler.d/nulldb.rb
+RUN foreman-rake apipie:cache:index RAILS_ENV=production DATABASE_URL=nulldb://nohost
+
+RUN mv /usr/share/foreman/db/schema.rb /usr/share/foreman/db/schema.rb.nulldb
+RUN rm -f /usr/share/foreman/bundler.d/nulldb.rb
+RUN yum -y remove tfm-rubygem-activerecord-nulldb-adapter && yum clean all
+# end generate apipie:cache:index
+
+{% if rpms_to_install | length > 0 %}
```

Comment:
It is possible with a Dockerfile to pass environment variables that are treated as arguments to the image build. That allows a standard Dockerfile to be flexible. Did you weigh that strategy, and avoiding the need for a build directory role while keeping the Dockerfile usable outside of Ansible against have a templated file? If so, what led you to this conclusion?

Reply:
I actually started with `docker build --build-arg FOREMAN_VERSION='x.y' .` but quickly transitioned to templating with Ansible as it allows me to achieve what I need more easily. While using ARG or ENV does allow for some customization it is fairly limited - for example passing arrays is not supported.

We could have something like the following for plugin rpms:

```
RUN if [ -n $RPMS_TO_INSTALL ]; then yum -y install $RPMS_TO_INSTALL; else echo "No plugins specified, skipping"; fi
```

Significant drawback is that we would need to pass any complex parameters as simple strings and place a logic to process them into a bash script. That does not mean we can't have Dockerfile without Ansible templating, we probably can. It would just be less readable and more work to make all the pieces fit, though I am not sure how I would do those custom snippets. So I would go with templating unless we have a requirement that prevents us using it.